alphabase/scoring/ml_scoring_base.py

```diff
@@ -0,0 +1,378 @@
+        df:pd.DataFrame,
+        fdr_level:str=None,
+        per_raw_fdr:bool=None,
+    )->pd.DataFrame:
+        if fdr_level is None: 
+            fdr_level = self.fdr_level
+        if per_raw_fdr is None: 
+            per_raw_fdr = self.per_raw_fdr
+
+        if per_raw_fdr:
+            return self._estimate_fdr_per_raw(
+                df, fdr_level=fdr_level
+            )
+
+        if fdr_level == 'psm':
+            return self._estimate_psm_fdr(df)
+        else:
+            if fdr_level == 'precursor':
+                _df = df.groupby([
+                    'sequence','mods','mod_sites','charge','decoy'
+                ])['ml_score'].max()
+            elif fdr_level == 'peptide':
+                _df = df.groupby([
+                    'sequence','mods','mod_sites','decoy'
+                ])['ml_score'].max()
+            else:
+                _df = df.groupby(['sequence','decoy'])['ml_score'].max()
+            _df = self._estimate_psm_fdr(_df)
+            df['fdr'] = fdr_from_ref(
+                df['ml_score'].values, _df['ml_score'].values, 
+                _df['fdr'].values
+            )
+        return df
+
+    def _train(self, 
+        train_t_df:pd.DataFrame, 
+        train_d_df:pd.DataFrame
+    ):
+        train_t_df = train_t_df[train_t_df.fdr<=self.training_fdr]
+
+        if len(train_t_df) > self.max_training_sample:
+            train_t_df = train_t_df.sample(
+                n=self.max_training_sample, 
+                random_state=1337
+            )
+        if len(train_d_df) > self.max_training_sample:
+            train_d_df = train_d_df.sample(
+                n=self.max_training_sample,
+                random_state=1337
+            )
+
+        train_df = pd.concat((train_t_df, train_d_df))
+        train_label = np.ones(len(train_df),dtype=np.int32)
+        train_label[len(train_t_df):] = 0
+
+        self._ml_model.fit(
+            train_df[self.feature_list].values, 
+            train_label
+        )
+
+    def _predict(self, test_df):
+        try:
+            test_df['ml_score'] = self._ml_model.decision_function(
+                test_df[self.feature_list].values
+            )
+        except AttributeError:
+            test_df['ml_score'] = self._ml_model.predict_proba(
+                test_df[self.feature_list].values
+            )
+        return test_df
+
+    def _train_and_score(self,
+        df:pd.DataFrame
+    )->pd.DataFrame:
+
+        df_target = df[df.decoy == 0]
+        df_decoy = df[df.decoy != 0]
+
+        if (
+            np.sum(df_target.fdr<=self.training_fdr) < 
+            self.min_training_sample or
+            len(df_decoy) < self.min_training_sample
+        ):
+            return df
+        
+        self._train(df_target, df_decoy)
+        test_df = pd.concat(
+            [df_target, df_decoy],
+            ignore_index=True
+        )
+    
+        return self._predict(test_df)
+
+    def _cv_score(self, df:pd.DataFrame)->pd.DataFrame:
```

Comment:
For totally random PSM selection, it is a better idea to use cross_val_score. 

Sometimes, I won't use cv_score as I would like to split the data by ourselves based on some columns like 'sequence' to prevent to put same sequences (diff PSMs) into these two training sets.